### ISSUE TITLE
ConnectionError when running a config file

### ISSUE THREAD

**Issue by surajsingireddy** [twosixlabs/armory#636]
Following the Getting Started guide on Windows 10, I encounter a ConnectionError when running an example scenario config file.

```
(base) C:\Users\suraj>conda create -n armory-test python=3.6
Collecting package metadata (current_repodata.json): done
Solving environment: done

## Package Plan ##

  environment location: C:\Users\suraj\AppData\Local\Continuum\miniconda3\envs\armory-test

  added / updated specs:
    - python=3.6


The following NEW packages will be INSTALLED:

  certifi            pkgs/main/win-64::certifi-2020.4.5.2-py36_0
  pip                pkgs/main/win-64::pip-20.1.1-py36_1
  python             pkgs/main/win-64::python-3.6.10-h9f7ef89_2
  setuptools         pkgs/main/win-64::setuptools-47.3.0-py36_0
  sqlite             pkgs/main/win-64::sqlite-3.31.1-h2a8f88b_1
  vc                 pkgs/main/win-64::vc-14.1-h0510ff6_4
  vs2015_runtime     pkgs/main/win-64::vs2015_runtime-14.16.27012-hf0eaf9b_2
  wheel              pkgs/main/win-64::wheel-0.34.2-py36_0
  wincertstore       pkgs/main/win-64::wincertstore-0.2-py36h7fe50ca_0
  zlib               pkgs/main/win-64::zlib-1.2.11-h62dcd97_4


Proceed ([y]/n)? y

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate armory-test
#
# To deactivate an active environment, use
#
#     $ conda deactivate


(base) C:\Users\suraj>conda activate armory-test

(armory-test) C:\Users\suraj>pip install armory-testbed
Collecting armory-testbed
  Using cached armory_testbed-0.9.0-py3-none-any.whl (119 kB)
Collecting docker
  Using cached docker-4.2.1-py2.py3-none-any.whl (143 kB)
Collecting jsonschema
  Using cached jsonschema-3.2.0-py2.py3-none-any.whl (56 kB)
Collecting coloredlogs
  Using cached coloredlogs-14.0-py2.py3-none-any.whl (43 kB)
Collecting pymongo
  Using cached pymongo-3.10.1-cp36-cp36m-win_amd64.whl (354 kB)
Collecting requests
  Using cached requests-2.23.0-py2.py3-none-any.whl (58 kB)
Collecting pypiwin32==223; sys_platform == "win32" and python_version >= "3.6"
  Using cached pypiwin32-223-py3-none-any.whl (1.7 kB)
Collecting websocket-client>=0.32.0
  Using cached websocket_client-0.57.0-py2.py3-none-any.whl (200 kB)
Collecting six>=1.4.0
  Using cached six-1.15.0-py2.py3-none-any.whl (10 kB)
Collecting importlib-metadata; python_version < "3.8"
  Using cached importlib_metadata-1.6.1-py2.py3-none-any.whl (31 kB)
Requirement already satisfied: setuptools in c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages (from jsonschema->armory-testbed) (47.3.0.post20200616)
Processing c:\users\suraj\appdata\local\pip\cache\wheels\d1\8a\1c\32ab9017418a2c64e4fbaf503c08648bed2f8eb311b869a464\pyrsistent-0.16.0-cp36-cp36m-win_amd64.whl
Collecting attrs>=17.4.0
  Using cached attrs-19.3.0-py2.py3-none-any.whl (39 kB)
Collecting humanfriendly>=7.1
  Using cached humanfriendly-8.2-py2.py3-none-any.whl (86 kB)
Collecting chardet<4,>=3.0.2
  Using cached chardet-3.0.4-py2.py3-none-any.whl (133 kB)
Collecting idna<3,>=2.5
  Using cached idna-2.9-py2.py3-none-any.whl (58 kB)
Requirement already satisfied: certifi>=2017.4.17 in c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages (from requests->armory-testbed) (2020.4.5.2)
Collecting urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1
  Using cached urllib3-1.25.9-py2.py3-none-any.whl (126 kB)
Collecting pywin32>=223
  Using cached pywin32-228-cp36-cp36m-win_amd64.whl (9.1 MB)
Collecting zipp>=0.5
  Using cached zipp-3.1.0-py3-none-any.whl (4.9 kB)
Processing c:\users\suraj\appdata\local\pip\cache\wheels\9c\d5\fa\c30bc1a04703cc6753dc9eea000865417a5d4d70331c77fc2e\pyreadline-2.1-py3-none-any.whl
Installing collected packages: pywin32, pypiwin32, six, websocket-client, chardet, idna, urllib3, requests, docker, zipp, importlib-metadata, pyrsistent, attrs, jsonschema, pyreadline, humanfriendly, coloredlogs, pymongo, armory-testbed
Successfully installed armory-testbed-0.9.0 attrs-19.3.0 chardet-3.0.4 coloredlogs-14.0 docker-4.2.1 humanfriendly-8.2 idna-2.9 importlib-metadata-1.6.1 jsonschema-3.2.0 pymongo-3.10.1 pypiwin32-223 pyreadline-2.1 pyrsistent-0.16.0 pywin32-228 requests-2.23.0 six-1.15.0 urllib3-1.25.9 websocket-client-0.57.0 zipp-3.1.0

(armory-test) C:\Users\suraj>git clone https://github.com/twosixlabs/armory-example.git
Cloning into 'armory-example'...
remote: Enumerating objects: 185, done.
remote: Counting objects: 100% (185/185), done.
remote: Compressing objects: 100% (121/121), done.
Receiving objects:  93% (459/493), 548.01 KiB | 340.00 KiB/s
Receiving objects: 100% (493/493), 608.28 KiB | 340.00 KiB/s, done.
Resolving deltas: 100% (302/302), done.
Checking connectivity... done.

(armory-test) C:\Users\suraj>cd armory-example

(armory-test) C:\Users\suraj\armory-example>armory run official_scenario_configs/cifar10_baseline.json
2020-06-17 12:45:19 LAPTOP-J0Q5A4OC armory.eval.evaluator[17372] ERROR Docker connection refused. Is Docker Daemon running?
Traceback (most recent call last):
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\connection.py", line 160, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\util\connection.py", line 84, in create_connection
    raise err
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\util\connection.py", line 74, in create_connection
    sock.connect(sa)
TimeoutError: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\connectionpool.py", line 677, in urlopen
    chunked=chunked,
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\connectionpool.py", line 381, in _make_request
    self._validate_conn(conn)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\connectionpool.py", line 976, in _validate_conn
    conn.connect()
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\connection.py", line 308, in connect
    conn = self._new_conn()
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\connection.py", line 172, in _new_conn
    self, "Failed to establish a new connection: %s" % e
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPSConnection object at 0x000002CAAF3C56A0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\requests\adapters.py", line 449, in send
    timeout=timeout
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\connectionpool.py", line 725, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\urllib3\util\retry.py", line 439, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='192.168.99.100', port=2376): Max retries exceeded with url: /v1.35/images/twosixarmory/pytorch:0.9.0/json (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002CAAF3C56A0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond',))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\suraj\AppData\Local\Continuum\miniconda3\envs\armory-test\Scripts\armory.exe\__main__.py", line 7, in <module>
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\__main__.py", line 564, in main
    func(sys.argv[2:], prog, description)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\__main__.py", line 229, in run
    rig = Evaluator(config, no_docker=args.no_docker)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\eval\evaluator.py", line 69, in __init__
    docker_client.images.get(kwargs["image_name"])
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\docker\models\images.py", line 316, in get
    return self.prepare_model(self.client.api.inspect_image(name))
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\docker\utils\decorators.py", line 19, in wrapped
    return f(self, resource_id, *args, **kwargs)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\docker\api\image.py", line 246, in inspect_image
    self._get(self._url("/images/{0}/json", image)), True
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\docker\utils\decorators.py", line 46, in inner
    return f(self, *args, **kwargs)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\docker\api\client.py", line 230, in _get
    return self.get(url, **self._set_request_timeout(kwargs))
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\requests\sessions.py", line 543, in get
    return self.request('GET', url, **kwargs)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\requests\sessions.py", line 530, in request
    resp = self.send(prep, **send_kwargs)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\requests\sessions.py", line 643, in send
    r = adapter.send(request, **kwargs)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\requests\adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='192.168.99.100', port=2376): Max retries exceeded with url: /v1.35/images/twosixarmory/pytorch:0.9.0/json (Caused by NewConnectionError('<urllib3.connection.HTTPSConnection object at 0x000002CAAF3C56A0>: Failed to establish a new connection: [WinError 10060] A connection attempt failed because the connected party did not properly respond after a period of time, or established connection failed because connected host has failed to respond',))
```

**Comment by davidslater** [comment]
The error is that the Docker connection was refused. Can you run
```
docker ps
```
and post the results?

Alternatively, can you run
```
armory run official_scenario_configs/cifar10_baseline.json --no-docker
```

**Comment by surajsingireddy** [comment]
Running ```docker ps``` revealed a similar issue. I had both Docker Desktop and Docker Toolbox installed, and after uninstalling Docker Toolbox and rerunning the config file I no longer encountered the ConnectionError. However, I then encountered the following issue, which appears to be caused by an invalid directory name.
```
2020-06-17 13:47:25 LAPTOP-J0Q5A4OC armory.eval.evaluator[12944] INFO Running evaluation script
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/conda/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.7/site-packages/armory/scenarios/base.py", line 223, in <module>
    run_config(args.config, args.from_file, args.check, args.mongo_host)
  File "/opt/conda/lib/python3.7/site-packages/armory/scenarios/base.py", line 176, in run_config
    _scenario_setup(config)
  File "/opt/conda/lib/python3.7/site-packages/armory/scenarios/base.py", line 152, in _scenario_setup
    os.makedirs(scenario_output_dir, exist_ok=True)
  File "/opt/conda/lib/python3.7/os.py", line 221, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 1] Operation not permitted: '/armory/outputs/2020-06-17T17:31:16.324901'

2020-06-17 13:47:42 LAPTOP-J0Q5A4OC armory.eval.evaluator[12944] WARNING Shutting down container
2020-06-17 13:47:42 LAPTOP-J0Q5A4OC armory.docker.management[12944] INFO Stopping instance: 5357aaa91d
2020-06-17 13:47:42 LAPTOP-J0Q5A4OC armory.eval.evaluator[12944] INFO Deleting tmp_dir C:\Users\suraj\.armory\tmp\2020-06-17T17:31:16.324901
2020-06-17 13:47:42 LAPTOP-J0Q5A4OC armory.eval.evaluator[12944] ERROR Error removing tmp_dir C:\Users\suraj\.armory\tmp\2020-06-17T17:31:16.324901
Traceback (most recent call last):
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\eval\evaluator.py", line 121, in _cleanup
    shutil.rmtree(self.tmp_dir)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\shutil.py", line 500, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\shutil.py", line 382, in _rmtree_unsafe
    onerror(os.listdir, path, sys.exc_info())
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\shutil.py", line 380, in _rmtree_unsafe
    names = os.listdir(path)
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\Users\\suraj\\.armory\\tmp\\2020-06-17T17:31:16.324901'
2020-06-17 13:47:42 LAPTOP-J0Q5A4OC armory.eval.evaluator[12944] INFO Removing output_dir C:\Users\suraj\.armory\outputs\2020-06-17T17:31:16.324901 if empty
```

**Comment by davidslater** [comment]
Interesting. For this error, are you running:
```
armory run official_scenario_configs/cifar10_baseline.json
```
?

**Comment by surajsingireddy** [comment]
Yes, I ran it again to be sure and got the same error.
```
(armory-test) C:\Users\suraj\armory-example>armory run official_scenario_configs/cifar10_baseline.json
2020-06-17 14:04:31 LAPTOP-J0Q5A4OC armory.docker.management[16624] INFO ARMORY Instance 2dd7849116 created.
2020-06-17 14:04:31 LAPTOP-J0Q5A4OC armory.eval.evaluator[16624] INFO Running evaluation script
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/opt/conda/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/opt/conda/lib/python3.7/site-packages/armory/scenarios/base.py", line 223, in <module>
    run_config(args.config, args.from_file, args.check, args.mongo_host)
  File "/opt/conda/lib/python3.7/site-packages/armory/scenarios/base.py", line 176, in run_config
    _scenario_setup(config)
  File "/opt/conda/lib/python3.7/site-packages/armory/scenarios/base.py", line 152, in _scenario_setup
    os.makedirs(scenario_output_dir, exist_ok=True)
  File "/opt/conda/lib/python3.7/os.py", line 221, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 1] Operation not permitted: '/armory/outputs/2020-06-17T18:04:31.006200'

2020-06-17 14:04:35 LAPTOP-J0Q5A4OC armory.eval.evaluator[16624] WARNING Shutting down container
2020-06-17 14:04:35 LAPTOP-J0Q5A4OC armory.docker.management[16624] INFO Stopping instance: 2dd7849116
2020-06-17 14:04:35 LAPTOP-J0Q5A4OC armory.eval.evaluator[16624] INFO Deleting tmp_dir C:\Users\suraj\.armory\tmp\2020-06-17T18:04:31.006200
2020-06-17 14:04:35 LAPTOP-J0Q5A4OC armory.eval.evaluator[16624] ERROR Error removing tmp_dir C:\Users\suraj\.armory\tmp\2020-06-17T18:04:31.006200
Traceback (most recent call last):
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\eval\evaluator.py", line 121, in _cleanup
    shutil.rmtree(self.tmp_dir)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\shutil.py", line 500, in rmtree
    return _rmtree_unsafe(path, onerror)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\shutil.py", line 382, in _rmtree_unsafe
    onerror(os.listdir, path, sys.exc_info())
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\shutil.py", line 380, in _rmtree_unsafe
    names = os.listdir(path)
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\Users\\suraj\\.armory\\tmp\\2020-06-17T18:04:31.006200'
2020-06-17 14:04:35 LAPTOP-J0Q5A4OC armory.eval.evaluator[16624] INFO Removing output_dir C:\Users\suraj\.armory\outputs\2020-06-17T18:04:31.006200 if empty
```

**Comment by davidslater** [comment]
It looks like a permission issue within the docker container.

Can you run:
```
armory run official_scenario_configs/cifar10_baseline.json --interactive
```
and then follow the `docker exec` command listed there?

Once inside the container, try the following command:
```
mkdir /armory/outputs/test
```

**Comment by surajsingireddy** [comment]
Unfortunately I ran into the same issue.
```
(armory-test) C:\Users\suraj\armory-example>armory run official_scenario_configs/cifar10_baseline.json --interactive
2020-06-17 14:40:12 LAPTOP-J0Q5A4OC armory.docker.management[18552] INFO ARMORY Instance 07a8a97d8d created.
2020-06-17 14:40:12 LAPTOP-J0Q5A4OC armory.eval.evaluator[18552] WARNING Shutting down container
2020-06-17 14:40:12 LAPTOP-J0Q5A4OC armory.docker.management[18552] INFO Stopping instance: 07a8a97d8d
Traceback (most recent call last):
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "C:\Users\suraj\AppData\Local\Continuum\miniconda3\envs\armory-test\Scripts\armory.exe\__main__.py", line 7, in <module>
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\__main__.py", line 564, in main
    func(sys.argv[2:], prog, description)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\__main__.py", line 234, in run
    check_run=args.check,
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\eval\evaluator.py", line 177, in run
    self._run_interactive_bash(runner)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\site-packages\armory\eval\evaluator.py", line 245, in _run_interactive_bash
    os.makedirs(tmp_dir)
  File "c:\users\suraj\appdata\local\continuum\miniconda3\envs\armory-test\lib\os.py", line 220, in makedirs
    mkdir(name, mode)
OSError: [WinError 123] The filename, directory name, or volume label syntax is incorrect: 'C:\\Users\\suraj\\.armory\\tmp\\2020-06-17T18:40:11.203445'
```

**Comment by davidslater** [comment]
That's my bad. Try:
Can you run:
```
armory launch pytorch --interactive
```
and then follow the docker exec command listed there?

Once inside the container, try the following command:
```
mkdir /armory/outputs/test
```

**Comment by surajsingireddy** [comment]
Worked until I tried making the folder inside the container.
```
(base) root@49144ce94c62:/workspace# mkdir /armory/outputs/test
mkdir: cannot create directory '/armory/outputs/test': Operation not permitted
(base) root@49144ce94c62:/workspace# ls
ls: reading directory '.': Operation not permitted
(base) root@49144ce94c62:/workspace# ls armory
ls: cannot access 'armory': No such file or directory
(base) root@49144ce94c62:/workspace# mkdir armory
mkdir: cannot create directory 'armory': Operation not permitted
```

**Comment by davidslater** [comment]
Adding @seanpmorgan as he has done more windows testing than I have. It looks like it's an issue with the read/write permissions you give to docker, but it may be a user issue.

@surajsingireddy can you write the docker exec command that armory output? I want to see what it gave as the uid mapping.

**Comment by surajsingireddy** [comment]
The command was 
`docker exec -it -u 0:0 49144ce94c bash`

**Comment by davidslater** [comment]
Should have been fixed by #639.
